perf_timer.py

```diff
@@ -476,22 +503,204 @@ def __exit__(self, excType, excVal, excTb):
 		:type output: None or :class:`collections.Callable` or str
+		
+		:param name: The name of the application. Running in normal python mode, this will default to the name of the __main__ module. This only affects the header in HTML mode.
+		:type name: None or str
 		"""
 
+		if name is None:
+			name = os.path.basename(sys.modules["__main__"].__file__)
+
+		if output is None:
+			if reportMode != ReportMode.HTML:
+				# pylint: disable=invalid-name,missing-docstring
+				def printIt(*args, **kwargs):
+					print(*args, **kwargs)
+
+				output = printIt
+			else:
+				output = os.path.basename(os.path.splitext(name)[0] + "_PERF.html")
+
+		elementsByFrame = {}
+		earliestByFrame = {}
+		latestByFrame = {}
+		allFramesQueue = deque()
+		allFramesAnnotations = deque()
+		annotationsByFrame = {}
+		while True:
+			try:
+				pair = PerfTimer.perfQueue.popleft()
+				frame = pair[4]
+				if frame in earliestByFrame:
+					earliestByFrame[frame] = min(earliestByFrame[frame], pair[5])
+					latestByFrame[frame] = max(latestByFrame[frame], pair[6])
+				else:
+					earliestByFrame[frame] = pair[5]
+					latestByFrame[frame] = pair[6]
+				if PerfTimer.minFrameTime is not None and (duration * 1000) < PerfTimer.minFrameTime:
+					continue
+				elementsByFrame.setdefault(frame, deque()).append(pair)
+				annotationsByFrame[frame] = deque();
+				allFramesQueue.append(pair)
+			except IndexError:
+				break
+
+		while True:
+			try:
+				pair = PerfTimer.annotations.popleft()
+				frame = pair[2]
+				globalEarliest = min(earliestByFrame.values())
+				allFramesPair = (pair[0], pair[1], pair[2], pair[3] - globalEarliest)
+				allFramesAnnotations.append(allFramesPair)
+				# Convert timestamp to a relative timestamp for this frame
+				pair = (pair[0], pair[1], pair[2], pair[3] - earliestByFrame[frame])
+				annotationsByFrame[frame].append(pair)
+			except IndexError:
+				break
+				
+		if len(elementsByFrame) > 1 and reportMode == ReportMode.HTML:
+			if not os.path.exists(os.path.join(os.path.dirname(output), "frames")):
+				os.mkdir(os.path.join(os.path.dirname(output), "frames"))
+			if __name__ == "__main__":
+				print("Generating combined frame output...")
+			PerfTimer.perfQueue = allFramesQueue
+			#PerfTimer.annotations = allFramesAnnotations
+			thisOutput = os.path.join(os.path.dirname(output), "frames", "_ALL.".join(os.path.basename(output).rsplit(".", 1)))
+			PerfTimer._printPerfReport(reportMode, thisOutput, None, name)
+
+		for key in sorted(elementsByFrame.keys()):
+			if key is not None:
+				if reportMode != ReportMode.HTML:
+					output("==============================")
+					output("Frame #{}".format(key))
+					output("==============================")
+				elif __name__ == "__main__":
+					sys.stdout.write("\rGenerating individual frame output for frame {}...".format(key))
```

Comment:
You have `\r` starting off this string.  Did you mean `\n`?

Reply:
I did not. `\r` is carriage return - return to beginning of line. `\n` moves the carat to the next line. This is overwriting the existing line so that only the number at the end changes and it doesn't spam the screen.